expression/builtin_string.go

```diff
@@ -200,8 +200,12 @@ func (c *concatFunctionClass) getFunction(args []Expression, ctx context.Context
 		return nil, errors.Trace(err)
 	}
 	for i := range args {
-		bf.tp.Flag |= mysql.BinaryFlag
-		bf.tp.Flen += args[i].GetType().Flen
+		argType := args[i].GetType()
+		if types.IsBinaryStr(argType) {
+			bf.tp.Flag |= mysql.BinaryFlag
```

Comment:
in the baseStringBuiltinFunc.getRetTp() will check the binary flag, and set charset and collation. @zz-jason 
```
func (b *baseStringBuiltinFunc) getRetTp() *types.FieldType {
    tp, flen := b.tp, b.tp.Flen
    if flen >= mysql.MaxBlobWidth {
        tp.Tp = mysql.TypeLongBlob
    } else if flen >= 65536 {
        tp.Tp = mysql.TypeMediumBlob
    }
    if mysql.HasBinaryFlag(tp.Flag) {
        tp.Charset, tp.Collate = charset.CharsetBin, charset.CollationBin
    } else {
        tp.Charset, tp.Collate = charset.CharsetUTF8, charset.CollationUTF8
    }
    return tp
}
```